url47fcfc0133_baliosian_tau_fst_tgz-pJ8-uy_edu_fing_mina_fsa_test_old_TestComposition1J8/src/uy/edu/fing/mina/lupa/LupaExporter.java

```diff
@@ -0,0 +1,826 @@
+				out = out.concat("events." + functionName + " = function(e) \n");
+//				out = out.concat(functionHeader(functionName) + "\n");
+				out = out.concat("\tlocal left= events." + functionName(leftTf) + "(e)\n");
+				out = out.concat("\tlocal right= events."	+ functionName(rightTf) + "(e)\n");
+				if (comp.op == Operator.AND) {
+					out = out.concat("\tif left < right then return left else return right end\n");
+				} else if (comp.op == Operator.OR) {
+					out = out.concat("\tif left > right then return left else return right end\n");
+
+				} 
+
+				out = out.concat("end\n");
+			}
+			else if (function.isNot()) {
+				out = out.concat("-- " + function.getName() + "\n");
+				out = out.concat("events." + functionName + " = function(e) \n");
+				TfI nonneg = function.not();
+				out = out.concat("\tlocal nonneg = events." + functionName(nonneg) + "(e)\n");
+				out = out.concat("\treturn 1-nonneg\n");
+				out = out.concat("end\n");
+			}
+			else{
+				throw new UnsupportedTFFSTException("A problematic function (" + function.getName() +") was detected. A generated EVENT function should always be composite or not.");
+			}
+		}
+		
+		
+		out = out.concat("-----------------------------------------------\n");
+		out = out.concat("--           END COMPOSITE EVENTS            --\n");
+		out = out.concat("-----------------------------------------------\n");
+		
+		return out;
+	}
+
+
+  /**
+   * Creates a String with actions functions that are CompositeTf or Nots.
+   * 
+   * @param tffst
+   *          The TFFST to export.
+   * @return The updated String objects with the new function bodies that it may
+   *         have.
+   * @throws UnsupportedTFFSTException
+   */
+  private static String compositeActionsToLupa(Tffst tffst) throws UnsupportedTFFSTException {
+    String out = new String();
+
+    out = out.concat("-----------------------------------------------\n");
+    out = out.concat("--          BEGIN COMPOSITE ACTIONS          --\n");
+    out = out.concat("-----------------------------------------------\n");
+    CompositeTf comp;
+    TfI leftTf;
+    TfI rightTf;
+    String functionName;
+    for (TfI function : compositeActions.values()) {
+      functionName = functionName(function);
+      if (function instanceof CompositeTf) {
+        comp = (CompositeTf) function;
+        leftTf = comp.left;
+        rightTf = comp.right;
+        out = out.concat("-- " + function.getName() + "\n");
+        // out = out.concat(functionHeader(functionName) + "\n");
+        out = out.concat("actions." + functionName + " = function(e)\n");
+        out = out.concat("\tlocal left= actions." + functionName(leftTf) + "(e)\n");
+        out = out.concat("\tlocal right= actions." + functionName(rightTf) + "(e)\n");
+        if (comp.op == Operator.AND) {
+          out = out.concat("\tif left < right then return left else return right end\n");
+        } else if (comp.op == Operator.OR) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An equality operator (== or !=) is used to compare strings, which matches strings based on address.

**Fix**: Do content comparison using the equals() method to compare the values of the strings.